webapp/src/components/github_milestone_selector/github_milestone_selector.jsx

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {PureComponent} from 'react';
+import PropTypes from 'prop-types';
+
+import IssueAttributeSelector from 'components/issue_attribute_selector';
+
+export default class GithubMilestoneSelector extends PureComponent {
+    static propTypes = {
+        repo: PropTypes.string.isRequired,
+        theme: PropTypes.object.isRequired,
+        selectedMilestone: PropTypes.number.isRequired,
+        onChange: PropTypes.func.isRequired,
+        milestones: PropTypes.array,
+        actions: PropTypes.shape({
+            getMilestones: PropTypes.func.isRequired,
+        }).isRequired,
+    };
+
+    loadMilestones = async () => {
+        if (this.props.repo === '') {
+            return [];
+        }
+
+        const options = await this.props.actions.getMilestones(this.props.repo);
+
+        if (options.error) {
+            throw new Error('Failed to load milestones');
+        }
+
+        if (!options || !options.data) {
+            return [];
+        }
+
+        return options.data.map((option) => ({
+            value: option.title,
+            label: option.title,
+        }));
+    };
+
+    onChange = (selection) => {
+        if (!this.props.milestones) {
+            this.props.onChange(0);
+            return;
+        }
+
+        // we have to find the selected milestone from the options in order to insert its number
+        const milestone = this.props.milestones.find((m) => m.title === selection);
```

Comment:
Can't we just pass the number as the option value?

Reply:
@larkox So let's take this milestone as an example: https://github.com/mattermost/mattermost-plugin-github/milestone/7

I think it's better to show the option for this milestone as `v1.2.0` rather than it's number, which is 7.

Reply:
I am not sure I am following, and I might not understand completely how React-Select works.
On a dropdown you have a "label" and a "value". The "value" is the "internal" value used for computation. The "label" is the user facing text.
If we set the label to "m.title", the user will see "v1.2.0". If then we set the value to milestone.number, we can work with the value we need (number). That would allow us to not use the redux store here and make the logic way more simple.
Does react-select works different as this?

Reply:
@larkox Your suggestion makes sense to me personally. 

However, as per @mickmister's earlier suggestion (https://github.com/mattermost/mattermost-plugin-github/pull/343#discussion_r466608271), we're only storing the milestone number in the parent component state.

So, in the `render` method of `GithubMilestoneSelector`, we have to loop through `this.props.milestones` in order to find the label for the selected milestone so that the label can be rendered in the widget rather than the number.

If we kept track of the milestone label in the parent component state as well as the milestone number, we wouldn't need to keep milestone options in redux state to loop through. But in that case, the way we store labels & assignees in the parent component state would be different than the way we store milestone. (We only store arrays of strings for labels & assignees in the parent component state.)

@larkox @mickmister Please let me know how we want to move forward with this. Personally, I'm fine with either option.

Reply:
I get what you are saying, but I would like to see milestone removed from the redux store.

To keep track of the label, we can store the whole milestone, or a simplified version with label and value. That would not conflict with `@mickmister`'s comment, I think, since setting this to null will mean "no selection", and no "truthy" value would be in place. But I am surprised that React-Select does not handle well just passing the value to the `value` prop, as would a normal html select component. Also, I believe that making the simplification of `label == value` is simplifying too much in issue attribute selector. Most of the times you want a human readable value for label, and a machine readable value for value.

@mickmister thoughts?

Reply:
@larkox I updated the PR as per your suggestion.

Reply:
When using react-select, it's common to just store the `value` in state, then do something like this in the `render` function. We need to tell it which "option" or "options" are selected.

```js
const selectedOption = this.state.options.find(opt => opt.value === this.state.value);
return (
	<ReactSelect
		options={this.state.options}
		value={selectedOption}
	/>
);
```